crypto/evp/pmeth_lib.c

```diff
@@ -816,76 +816,37 @@ int EVP_PKEY_CTX_ctrl_uint64(EVP_PKEY_CTX *ctx, int keytype, int optype,
+        name = OSSL_ASYM_CIPHER_PARAM_OAEP_DIGEST;
+    else if (strcmp(name, "rsa_oaep_label") == 0)
+        name = OSSL_ASYM_CIPHER_PARAM_OAEP_LABEL;
 # ifndef OPENSSL_NO_DH
-    if (strcmp(name, "dh_pad") == 0) {
-        int pad;
-
-        pad = atoi(value);
-        return EVP_PKEY_CTX_set_dh_pad(ctx, pad);
-    }
+    else if (strcmp(name, "dh_pad") == 0)
+        name = OSSL_EXCHANGE_PARAM_PAD;
 # endif
-    if (strcmp(name, "digest") == 0) {
-        int ret;
-        EVP_MD *md;
-
-        if (!EVP_PKEY_CTX_IS_SIGNATURE_OP(ctx) || ctx->op.sig.signature == NULL)
-            return 0;
-        md = EVP_MD_fetch(ossl_provider_library_context(ctx->op.sig.signature->prov),
-                          value, NULL);
-        if (md == NULL)
-            return 0;
-        ret = EVP_PKEY_CTX_set_signature_md(ctx, md);
-        EVP_MD_meth_free(md);
-        return ret;
-    }
-
-    if (strcmp(name, "rsa_padding_mode") == 0) {
-        int pm;
-
-        if (strcmp(value, "pkcs1") == 0) {
-            pm = RSA_PKCS1_PADDING;
-        } else if (strcmp(value, "sslv23") == 0) {
-            pm = RSA_SSLV23_PADDING;
-        } else if (strcmp(value, "none") == 0) {
-            pm = RSA_NO_PADDING;
-        } else if (strcmp(value, "oeap") == 0) {
-            pm = RSA_PKCS1_OAEP_PADDING;
-        } else if (strcmp(value, "oaep") == 0) {
-            pm = RSA_PKCS1_OAEP_PADDING;
-        } else if (strcmp(value, "x931") == 0) {
-            pm = RSA_X931_PADDING;
-        } else if (strcmp(value, "pss") == 0) {
-            pm = RSA_PKCS1_PSS_PADDING;
-        } else {
-            ERR_raise(ERR_LIB_RSA, RSA_R_UNKNOWN_PADDING_TYPE);
-            return -2;
-        }
-        return EVP_PKEY_CTX_set_rsa_padding(ctx, pm);
-    }
 
-    if (strcmp(name, "rsa_mgf1_md") == 0)
-        return EVP_PKEY_CTX_set_rsa_mgf1_md_name(ctx, value, NULL);
-
-    if (strcmp(name, "rsa_oaep_md") == 0)
-        return EVP_PKEY_CTX_set_rsa_oaep_md_name(ctx, value, NULL);
-
-    if (strcmp(name, "rsa_oaep_label") == 0) {
-        unsigned char *lab;
-        long lablen;
-        int ret;
+    {
+        /*
+         * TODO(3.0) reduce the code above to only translate known legacy
+         * string to the corresponding core name (see core_names.h), but
+         * otherwise leave it to this code block to do the actual work.
+         */
+        const OSSL_PARAM *settable = EVP_PKEY_CTX_settable_params(ctx);
+        OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+        int rv = 0;
 
-        lab = OPENSSL_hexstr2buf(value, &lablen);
-        if (lab == NULL)
+        if (!OSSL_PARAM_allocate_from_text(&params[0], settable, name, value,
+                                           strlen(value)))
             return 0;
-        ret = EVP_PKEY_CTX_set0_rsa_oaep_label(ctx, lab, lablen);
-        if (ret <= 0)
-            OPENSSL_free(lab);
-        return ret;
+        if (EVP_PKEY_CTX_set_params(ctx, params))
+            rv = 1;
```

Comment:
As it is this is not flexible enough:
- for some params we must make sure that the provider supports them as settable as ignoring a param in that case would be breaking existing legacy behavior and open security risks: e.g., `name == "ecdh_cofactor_mode"` if the provider does not support this parameter as settable and does not react on this request this is not signaled to the application with an error (i.e., breaking existing behavior) and the ECDH computation will be performed without the cofactor-tweak,  exposing to serious small subgroups attacks! (i.e. security risk)
- for some params we used to raise specific errors in specific cases: e.g., `name == ecdh_cofactor_md` we pass the name of the digest algorithm and we used to raise `EC_R_INVALID_DIGEST` if `EVP_get_digestbyname(value) == NULL`. This might be a minor concern but going directly as a param for the provider we add a useless round-trip (as we would expect the provider `ctx_set_param` implementations to try and go back into core to fetch the matching digest).

The problem of ignoring by default unknown parameters on a `set_params` (or `get_params`) and breaking existing behavior actually applies to both my examples (and I suspect also to the cases that this PR is already supporting.

If we tried to explicitly tweak an operation by setting a non-default digest, some special padding mode or a specific mode of operation, etc., we can't return success to the application (or `libcrypto`/`libssl`) if this has not been received correctly by the provider: this is required to not break existing behavior, but also to avoid undetected unsafe behavior!

Reply:
One more point that I forgot to mention before: some values might be received as strings by this function but might need conversion as the OSSL_PARAM type is not string (or is a string but not the value passed in as a string in the legacy API).

One example is once more `ecdh_cofactor_mode`. In this function the value is a string, but the OSSL_PARAM is an int that must be either -1, 0 or 1.

Reply:
Before making assumptions on strings vs numbers, you might want to read up on `OSSL_PARAM_allocate_from_text()`

https://github.com/openssl/openssl/blob/master/doc/man3/OSSL_PARAM_construct_from_text.pod

Reply:
Also, I believe that this same function answers your concern about finding out if a parameter is actually settable.

Reply:
I apologize for missing that function!

I can't do a proper review at the moment, the important thing for me is that the concerns above are indeed addressed.

Reply:
> I apologize for missing that function!

No problem.  There's so many new things that we're all prone to miss a few...